### PR TITLE
Log collection with PHP

### PR DESCRIPTION
**Pull request by l0k0ms**
### What does this PR do?

Add PHP log collection documentation.

### Motivation

It was missing

### Preview link

https://docs-staging.datadoghq.com/gus/log-php-collection/logs/languages/php/

### CODE REVIEW

File: content/logs/languages/php.md [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+  tag: "Documentation"
+  text: Learn how to explore your logs
+- link: /logs/faq/log-collection-troubleshooting-guide
+  tag: "FAQ"
+  text: Log Collection Troubleshooting Guide
+---
+
+## Overview
+
+Write your PHP logs into a file, then [use the Agent](/logs) to forward them to Datadog, choose between the following libraries:
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: Should this `[use the Agent](/logs)` link to `/agent` instead?

> `[use the Agent](/agent)`

File: content/logs/languages/php.md [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+    // load Monolog library
+    use Monolog\Logger;
+    use Monolog\Handler\StreamHandler;
+    use Monolog\Formatter\JsonFormatter;    
+    ```
+
+
+### Setup - Log into a file with Monolog
+
+The following configuration enables the Json formatting and writes the logs and events into the `application-json.log` file. Edit your code, right after the initialization of the Monolog instance and add a new handler:
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: I would make `Json` all uppercase.

> The following configuration enables the JSON formatting and writes the logs and events into the `application-json.log` file. Edit your code, right after the initialization of the Monolog instance and add a new handler:

File: content/logs/languages/php.md [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+$stream->setFormatter($formatter);
+
+// bind
+$log->pushHandler($stream);
+
+// an example
+$log->info('Adding a new user', array('username' => 'Seldaek'));
+```
+
+Then [stream your log file to Datadog](/logs/#custom-log-collection)
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: Should there be a period after this sentence?

> `Then [stream your log file to Datadog](/logs/#custom-log-collection).`

File: content/logs/languages/php.md [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+
+### Add meta field and context
+
+Much of the usefulness information comes from additional context data that you add to your logs and events.
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: I would rephrase this as:

> Much of the useful information comes from additional context data you can add to your logs and events.

File: content/logs/languages/php.md [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+// an example
+$log->info('Adding a new user', array('username' => 'Seldaek'));
+```
+
+Then [stream your log file to Datadog](/logs/#custom-log-collection)
+
+### Add meta field and context
+
+Much of the usefulness information comes from additional context data that you add to your logs and events.  
+Monolog makes this very convenient by providing methods to set thread local context data that is then submitted automatically with all events. At any moment, log an event with a contextual data:
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: I would rephrase this as:

> Monolog makes this convenient by providing methods for setting thread-local context data that is then submitted automatically with all events. At any moment, log an event with contextual data:

File: content/logs/languages/php.md [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+    // Add various generic context
+    $record['extra']['key'] = 'value';
+
+    return $record;
+});
+```
+
+### Framework integration 
+
+Monolog is a part of the following framework:
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: > Monolog is a part of the following frameworks:

File: content/logs/languages/php.md [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+
+Then [stream your log file to Datadog](/logs/#custom-log-collection)
+
+### Add meta field and context
+
+Much of the usefulness information comes from additional context data that you add to your logs and events.  
+Monolog makes this very convenient by providing methods to set thread local context data that is then submitted automatically with all events. At any moment, log an event with a contextual data:
+
+```php
+ <?php
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: For the code snippets that are one-liners, I would omit the `<?php` at the top - since it implies it's at the beginning of the file. Instead,  I would write:

> ```php
> $logger->info('Adding a new user', array('username' => 'Seldaek'));
> ```

File: content/logs/languages/php.md [hunk 8]
```diff
@@ -0,0 +1,503 @@
...
+
+$logger->info('Adding a new user', array('username' => 'Seldaek'));
+```
+
+But, most important, Monolog comes with a pre-processor feature. It's a simple callback enriches all your events with, for instance, the session id, the request id, and so on:
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: Instead of swaying the reader's opinion about which feature is the most important, I would simply describe the feature, and let the reader come to their own conclusion.

I would rephrase this as:

> Monolog comes with a _pre-processor_ feature. It's a simple callback that enriches your events with metadata you can set (e.g., the session id, the request id, etc.):

File: content/logs/languages/php.md [hunk 9]
```diff
@@ -0,0 +1,503 @@
...
+
+```php
+ <?php 
+    
+$log->pushProcessor(function ($record) {
+
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: I would delete this extra newline here.

File: content/logs/languages/php.md [hunk 10]
```diff
@@ -0,0 +1,503 @@
...
+$monolog = ...
+  
+///// Log shipper configuration
+
+$formatter = new JsonFormatter();
+$stream = new StreamHandler(__DIR__.'/application-json.log', Logger::DEBUG);
+$stream->setFormatter($formatter);
+
+$monolog->pushHandler($stream);
+return $r;});
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: Remove the `});` at the end of the `return` line

> `return $r;`

File: content/logs/languages/php.md [hunk 11]
```diff
@@ -0,0 +1,503 @@
...
+            channels: ~
+            # use json formatter
+            formatter: monolog.json_formatter
+            # logs alls events (debug trought fatal)
+            level: debug
+```
+
+#### PPI
+
+Look at your configuration directory, and edit `config_dev.yml` or `config_prod.yml` depending on your needs:
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: I would rephrase this to:

> In your configuration directory `/path/to/config/directory/`, you may edit the `config_dev.yml` and `config_prod.yml` to configure log management to suit your needs for both development and production environments.

File: content/logs/languages/php.md [hunk 12]
```diff
@@ -0,0 +1,503 @@
...
+
+$monolog->pushHandler($stream);
+return $r;});
+```
+
+and then [stream your log files to Datadog](/logs/#custom-log-collection)
+
+#### Symfony (v2+, v3+)
+
+Look at your configuration directory, and edit `config_dev.yml` or `config_prod.yml` depending on your needs:
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: I would rephrase this to:

> In your configuration directory `/path/to/config/directory/`, you may edit the `config_dev.yml` and `config_prod.yml` to configure log management to suit your needs for development and production environments.

File: content/logs/languages/php.md [hunk 13]
```diff
@@ -0,0 +1,503 @@
...
+
+```json
+{
+    "require": {
+        "cakephp/monolog": "*"
+    }
+}
+```
+
+Then, start by creating a logging configuration  file (i.e. `app/Config/log.php`) that you include early in your `app/Config/bootstrap.php`:
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: Two comments here:

1) There's an extra space between `configuration  file`
2) I'd remove the word `early`

> Then, start by creating a logging configuration file (i.e. `app/Config/log.php`) that you include in your `app/Config/bootstrap.php`:

File: content/logs/languages/php.md (hunk 1 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: Nop because agent configuration instructions for log collection are in `/logs`

File: content/logs/languages/php.md (hunk 2 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 3 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 4 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 5 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 6 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 7 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 8 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 9 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 10 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 11 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 12 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 13 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: 👌

File: content/logs/languages/php.md (hunk 1 above)
Comment (AndrewMcBurney) [reply content/logs/languages/php.md]: Ahh, gotcha - thanks Gus! :)

File: content/logs/languages/php.md [hunk 14]
```diff
@@ -0,0 +1,562 @@
...
+### Configure your Datadog agent
+
+Create a `php.d/conf.yaml` file in your `conf.d/` folder with the following content:
+
+```yaml
+init_config:
+
+instances:
+    
+##Log section
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: Could we add a space in between the `##`? It's a bit of a nitpick, but I think it'll make it a little bit more readable.

File: content/logs/languages/php.md [hunk 15]
```diff
@@ -0,0 +1,562 @@
...
+logs:
+
+    ## - type : file (mandatory) type of log input source (tcp / udp / file)
+    ##   port / path : (mandatory) Set port if type is tcp or udp. Set path if type is file
+    ##   service : (mandatory) name of the service owning the log
+    ##   source : (mandatory) attribute that defines which integration is sending the logs
+    ##   sourcecategory : (optional) Multiple value attribute. Can be used to refine the source attribtue
+    ##   tags: (optional) add tags to each logs collected
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: There's a space between the colon on each of the lines, with the exception of line with `tags:`. Personally, I like the format: `name: ` where there's no space between the name and the colon, but a space following. What do you think?

```yaml
    ## - type: file (mandatory) type of log input source (tcp / udp / file)
    ##   port/path: (mandatory) Set port if type is tcp or udp. Set path if type is file
    ##   service: (mandatory) name of the service owning the log
    ##   source: (mandatory) attribute that defines which integration is sending the logs
    ##   sourcecategory: (optional) Multiple value attribute. Can be used to refine the source attribtue
    ##   tags: (optional) add tags to each logs collected
```

File: content/logs/languages/php.md [hunk 16]
```diff
@@ -0,0 +1,562 @@
...
+  - type: file
+    path: /path/to/your/php/application-json.log
+    service: php
+    source: php
+    sourcecategory: sourcecode
+```
+
+### Add meta field and context
+
+Much of the useful information comes from additional context data added to your logs and events.
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: I would rephrase this line as:

> It's useful to add additional context data to your logs and events.

File: content/logs/languages/php.md [hunk 17]
```diff
@@ -0,0 +1,562 @@
...
+    
+##Log section
+logs:
+
+    ## - type : file (mandatory) type of log input source (tcp / udp / file)
+    ##   port / path : (mandatory) Set port if type is tcp or udp. Set path if type is file
+    ##   service : (mandatory) name of the service owning the log
+    ##   source : (mandatory) attribute that defines which integration is sending the logs
+    ##   sourcecategory : (optional) Multiple value attribute. Can be used to refine the source attribtue
+    ##   tags: (optional) add tags to each logs collected
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: [Same comment](https://github.com/DataDog/documentation/pull/2127/files#r175451646) here about the colon spacing.

File: content/logs/languages/php.md [hunk 18]
```diff
@@ -0,0 +1,562 @@
...
+    source: php
+    sourcecategory: sourcecode
+```
+
+#### Symfony (v2+, v3+)
+
+In your configuration directory `/path/to/config/directory/`, edit the `config_dev.yml` and `config_prod.yml` to configure log management to suit your needs for development and production environments.
+
+```yaml
+ # app/config/config.yml
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: Should there be an extra space proceeding this comment?

> ```yaml
> # app/config/config.yml
> monolog:
> ```

File: content/logs/languages/php.md [hunk 19]
```diff
@@ -0,0 +1,562 @@
...
+#            tags:
+#               - { name: monolog.processor, method: processRecord }
+          
+          
+    json_formatter:
+        class: Monolog\Formatter\JsonFormatter
+          
+    handlers:
+    
+            # Log shipper configuration
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: This comment is indented a bit more than the following `to_json_files` line.

File: content/logs/languages/php.md [hunk 20]
```diff
@@ -0,0 +1,562 @@
...
+```
+
+#### Laravel
+
+```php
+//file: bootstrap/app.php
+$app->configureMonologUsing(function($monolog) {
+    $monolog->pushHandler(...);
+    
+  // configure your logger as below
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: Instead of saying:

> `// configure your logger as below`

I would remove the `as`:

> `// configure your logger below`

File: content/logs/languages/php.md [hunk 21]
```diff
@@ -0,0 +1,562 @@
...
+return $app;
+```
+
+#### Silex
+
+```php
+// file: bootstrap
+$app->extend('monolog', function($monolog, $app) {
+    $monolog->pushHandler(...);
+    // configure your logger as below
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: [Same here](https://github.com/DataDog/documentation/pull/2127/files#r175455257).

File: content/logs/languages/php.md [hunk 22]
```diff
@@ -0,0 +1,562 @@
...
+});
+```
+
+#### Lumen
+
+```php
+//file: bootstrap/app.php
+$app->configureMonologUsing(function($monolog) {
+    $monolog->pushHandler(...);
+    // configure your logger as below
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: [Same comment here](https://github.com/DataDog/documentation/pull/2127/files#r175455257).

File: content/logs/languages/php.md [hunk 23]
```diff
@@ -0,0 +1,562 @@
...
+
+```json
+{
+    "require": {
+        "cakephp/monolog": "*"
+    }
+}
+```
+
+Then, start by creating a logging configuration file (i.e. `app/Config/log.php`) that you include your `app/Config/bootstrap.php`:
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: I would pluralize `include`:

> Then, start by creating a logging configuration file (i.e., `app/Config/log.php`) that you includes your `app/Config/bootstrap.php`:

File: content/logs/languages/php.md [hunk 24]
```diff
@@ -0,0 +1,562 @@
...
+
+1. Declare a Monolog JSON formatter as a service:
+
+    ```yaml
+    services:
+        monolog.json_formatter:
+            class: Monolog\Formatter\JsonFormatter
+    ```
+
+2. Wire the formatter in your Monolog configuration: declare the formatter field as follows:
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: What does `Wire` refer to? Should it be `Configure` instead?

File: content/logs/languages/php.md [hunk 25]
```diff
@@ -0,0 +1,562 @@
...
+    
+##Log section
+logs:
+
+    ## - type : file (mandatory) type of log input source (tcp / udp / file)
+    ##   port / path : (mandatory) Set port if type is tcp or udp. Set path if type is file
+    ##   service : (mandatory) name of the service owning the log
+    ##   source : (mandatory) attribute that defines which integration is sending the logs
+    ##   sourcecategory : (optional) Multiple value attribute. Can be used to refine the source attribtue
+    ##   tags: (optional) add tags to each logs collected
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: [Same comment](https://github.com/DataDog/documentation/pull/2127/files#r175451646) here about the colon spacing.

File: content/logs/languages/php.md [hunk 26]
```diff
@@ -0,0 +1,562 @@
...
+
+    ```php
+     <?php
+
+    namespace Acme\Bundle\MonologBundle\Log;
+
+    use Symfony\Component\HttpFoundation\Session\Session;
+
+    class SessionRequestProcessor
+    {
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: In previous PHP code samples, you have the brace on the same line as the `class`/`function` definitions. I would change this code sample to be as follows:

```php
     <?php

    namespace Acme\Bundle\MonologBundle\Log;

    use Symfony\Component\HttpFoundation\Session\Session;

    class SessionRequestProcessor {
      private $session;
      private $sessionId;
      private $requestId;
      private $_server;
      private $_get;
      private $_post;

      public function __construct(Session $session) {
        $this->session = $session;
      }

      public function processRecord(array $record) {
        if (null === $this->requestId) {
          if ('cli' === php_sapi_name()) {
            $this->sessionId = getmypid();
          } else {
            try {
              $this->session->start();
              $this->sessionId = $this->session->getId();
            } catch (\RuntimeException $e) {
              $this->sessionId = '????????';
            }
          }
          $this->requestId = substr(uniqid(), -8);
          $this->_server = array(
            'http.url' => (@$_SERVER['HTTP_HOST']).'/'.(@$_SERVER['REQUEST_URI']),
            'http.method' => @$_SERVER['REQUEST_METHOD'],
            'http.useragent' => @$_SERVER['HTTP_USER_AGENT'],
            'http.referer' => @$_SERVER['HTTP_REFERER'],
            'http.x_forwarded_for' => @$_SERVER['HTTP_X_FORWARDED_FOR']
          );
          $this->_post = $this->clean($_POST);
          $this->_get = $this->clean($_GET);
        }
        $record['http.request_id'] = $this->requestId;
        $record['http.session_id'] = $this->sessionId;
        $record['http.url'] = $this->_server['http.url'];
        $record['http.method'] = $this->_server['http.method'];
        $record['http.useragent'] = $this->_server['http.useragent'];
        $record['http.referer'] = $this->_server['http.referer'];
        $record['http.x_forwarded_for'] = $this->_server['http.x_forwarded_for'];

        return $record;
      }

      protected function clean($array) {
        $toReturn = array();
        foreach(array_keys($array) as $key) {
          if (false !== strpos($key, 'password')) {
            // Do not add
          } else if (false !== strpos($key, 'csrf_token')) {
            // Do not add
          } else {
            $toReturn[$key] = $array[$key];
          }
        }

        return $toReturn;
      }
    }
```

File: content/logs/languages/php.md [hunk 27]
```diff
@@ -0,0 +1,562 @@
...
+// bind
+$logger->addWriter($writer);
+Zend\Log\Logger::registerErrorHandler($logger);
+```
+
+Then [Stream your log files to Datadog](/logs/#custom-log-collection)
+
+### Add meta field and context
+
+Much of the usefulness information comes from additional context data that you can add to your logs and your events. Zend-Log makes this very convenient by providing methods to set thread local context data that is then submitted automatically with all events. At any moment, log an event with a contextual data:
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: I would rephrase as:

> `Much of the useful information comes from additional context data that you can add to your logs and your events.`

Comment (AndrewMcBurney) [reply content/logs/languages/php.md]: I would rephrase the last sentence as:

> `At any moment, log an event with contextual data:`

File: content/logs/languages/php.md [hunk 28]
```diff
@@ -0,0 +1,562 @@
...
+
+```php
+$logger->info('Adding a new user', array('username' => 'Seldaek'));
+```
+
+But, most important, the library comes with a processor feature. Processors allow you to provide additional information to logs in an automated fashion. They are called from the logger before the event is passed to the writers; they receive the event array, and return an event array on completion.
+
+Use cases include:
+
+* Providing exception backtrace information.
```
Comment (AndrewMcBurney) [inline content/logs/languages/php.md]: These first two bullet points have a period, but the last one does not. I prefer having no periods after bullet points, so I'd remove the periods from the first two.

File: content/logs/languages/php.md (hunk 14 above)
Comment (l0k0ms) [reply content/logs/languages/php.md]: it's like this in the default configuration files (it bothers me too.. but it's just the way it goes. :/)
